Added OTEL access logging envoy extension (#18336)
hashicorp/consul: website/content/docs/connect/proxies/envoy-extensions/configuration/otel-access-logging.mdx

      - [`BufferSizeBytes`](#arguments-config-buffersizebytes): number
      - [`FilterStateObjectsToLog`](#arguments-config-filterstateobjectstolog): list of strings
      - [`RetryPolicy`](#arguments-config-retrypolicy): map
        - [`RetryBackOff`](#arguments-config-retrypolicy-retrybackoff): map
          - [`BaseInterval`](#arguments-config-retrypolicy-retrybackoff): string | `1s`
          - [`MaxInterval`](#arguments-config-retrypolicy-retrybackoff): string | `30s`
        - [`NumRetries`](#arguments-config-retrypolicy-numretries): number
      - [`Body`](#arguments-config-body): string, number, boolean or list of bytes
      - [`Attributes`](#arguments-config-attributes): map of string to string, number, boolean or list of bytes
      - [`ResourceAttributes`](#arguments-config-resourceattributes): map of string to string, number, boolean or list of bytes

## Complete configuration

When each field is defined, an `otel-access-logging` configuration has the following form:

```hcl
Name = "builtin/otel-access-logging" 
Arguments = {
   ProxyType = "connect-proxy"
   ListenerType = "<inbound or outbound>"
   Config = {

>       LogName = "<The friendly name of the access log>"

Reviewer: ```suggestion
      LogName = "<user-readable name of the access log>"
```
As opposed to snarky name? :)

Specifies the name of the extension. Must be set to `builtin/otel-access-logging`.

#### Values

- Default: None
- This field is required.
- Data type: String value set to `builtin/otel-access-logging`.

### `Arguments`

Contains the global configuration for the extension.

#### Values

- Default: None
- This field is required.
- Data type: Map

### `Arguments.ProxyType`

Specifies the type of Envoy proxy that this extension applies to. The extension only applies to proxies that match this type and is ignored for all other proxy types. The only supported value is `connect-proxy`.

#### Values

- Default: `connect-proxy`
- This field is required.
- Data type: String

### `Arguments.ListenerType`

Specifies the type of listener the extension applies to. The listener type is either `inbound` or `outbound`. If the listener type is set to `inbound`, Consul applies the extension so the access logging is enabled when other services in the mesh send messages to the service attached to the proxy. If the listener type is set to `outbound`, Consul applies the extension so the access logging is enabled when the attached proxy sends messages to other services in the mesh.

#### Values

> - Default: inbound

Reviewer: ```suggestion
- Default: `inbound`
```

- This field is required.
- Data type is one of the following string values:
  - `inbound`
  - `outbound`

### `Arguments.Config`

Contains the configuration settings for the extension.

#### Values

- Default: None
- This field is required.
- Data type: Map

### `Arguments.Config.LogName`

> Specifies the friendly name of the access log to be returned in StreamAccessLogsMessage.Identifier. This allows the access log server to differentiate between different access logs coming from the same Envoy.

Reviewer: ```suggestion
Specifies the user-readable name of the access log to be returned in `StreamAccessLogsMessage.Identifier`. This allows the access log server to differentiate between different access logs coming from the same Envoy.
```

#### Values

> - Default: value given to ListenerType

Reviewer: ```suggestion
- Default: Inherits the value given to ListenerType
```
We're still firming this up so I'm leaving this for now, but I think we're going to standardize on "None" as the default even when the parameter inherits a value from another configuration. This is to differentiate from true default values programmed into the software that you must explicitly override. We'll describe inheritance in the description of the parameter. Does that make sense?

Author: Used the suggested way. Please check if the updated one sounds good.

### `Arguments.Config.GrpcService.Target.Timeout`

Specifies the maximum duration that a response can take to arrive upon request.
 
#### Values

- Default: `1s`
- Data type: String

### `Arguments.Config.GrpcService.Authority`

Specifies the authority header to send in the gRPC request. If this field is not set, the authority field is set to the cluster name. This field does not override the SNI that Envoy sends to the OpenTelemetry collector service.
 
#### Values

- Default: Cluster name
- Data type: String

### `Arguments.Config.GrpcService.InitialMetadata`

Specifies additional metadata to include in streams initiated to the `GrpcService`. You can specify metadata for injecting additional ad-hoc authorization headers, for example, `x-foo-bar: baz-key`. For more information, including details on header value syntax, refer to the [Envoy documentation on custom request headers](https://www.envoyproxy.io/docs/envoy/latest/configuration/http/http_conn_man/headers#config-http-conn-man-headers-custom-request-headers).
 
#### Values

- Default: None
- Data type: List of one or more key-value pairs:

   - KEY: String
   - VALUE: String

### `Arguments.Config.BufferFlushInterval`

> Specifies interval for flushing access logs to the gRPC stream. Logger will flush requests every time this interval is elapsed, or when batch size limit is hit, whichever comes first.

Reviewer: ```suggestion
Specifies an interval for flushing access logs to the gRPC stream. The logger flushes requests at the end of every interval or when the log reaches the batch size limit, whichever comes first.
```
Present tense and active voice

#### Values

- Default: `1s`
- Data type: String

### `Arguments.Config.BufferSizeBytes`

> Specifies soft size limit in bytes for access log entries buffer. Logger will buffer requests until this limit it hit, or every time flush interval is elapsed, whichever comes first. Setting it to zero effectively disables the batching.

Reviewer: ```suggestion
Specifies the soft size limit in bytes for the access log entries buffer. The logger buffers requests until it reaches this limit or every time the flush interval elapses, whichever comes first. Set this field to `0` to disable batching.
```

#### Values

- Default: `16384`
- Data type: Integer

### `Arguments.Config.FilterStateObjectsToLog`

> Specifies additional filter state objects to log in filter_state_objects. Logger will call FilterState::Object::serializeAsProto to serialize the filter state object.

Reviewer: ```suggestion
Specifies additional filter state objects to log in `filter_state_objects`. The logger calls `FilterState::Object::serializeAsProto` to serialize the filter state object.
```

| `BaseInterval` | Specifies the base interval for determining the next backoff computation. Set a value greater than `0` and less than or equal to the `MaxInterval` value.  | String    | `1s`    |
| `MaxInterval`  | Specifies the maximum interval between retries. Set the value greater than or equal to the `BaseInterval` value.                                           | String    | `10s`   |

### `Arguments.Config.RetryPolicy.NumRetries`

Specifies the number of times Envoy retries to fetch plugin data if the initial attempt is unsuccessful.

#### Values

- Default: `1`
- Data type: Integer

### `Arguments.Config.Body`

Specifies OpenTelemetry [LogResource](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto) fields, following [Envoy access logging formatting](https://www.envoyproxy.io/docs/envoy/latest/configuration/observability/access_log/usage). See ‘body’ in the LogResource proto for more details.

#### Values

- Default: None
- Data type: String

### `Arguments.Config.Attributes`

> Specifies `attributes` in the [LogResource](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto). See ‘attributes’ in the LogResource proto for more details.

Reviewer: ```suggestion
Specifies `attributes` in the [LogResource](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto). Refer to `attributes` in the LogResource proto for more details.
```
I'm a little unclear on what we're referring people to in the second sentence. Are we telling people to check an entry called `attributes` in some output? Would this be clearer to the target audience?

#### Values

- Default: None
- Data type: Map

### `Arguments.Config.ResourceAttributes`

> Specifies OpenTelemetry [Resource](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto#L51) attributes are filled with Envoy node info.

Reviewer: ```suggestion
Specifies OpenTelemetry [Resource](https://github.com/open-telemetry/opentelemetry-proto/blob/main/opentelemetry/proto/logs/v1/logs.proto#L51) attributes are filled with Envoy node information.
```
